az-cvm-vtpm/az-tdx-vtpm/src/hcl.rs

```diff
@@ -0,0 +1,192 @@
+    #[error("AkPub not found")]
+    AkPubNotFound,
+    #[error("parse error")]
+    ParseError(#[from] bincode::Error),
+}
+
+#[derive(Deserialize, Debug)]
+struct VarDataKeys {
+    keys: Vec<JsonWebKey>,
+}
+
+#[repr(u32)]
+#[derive(Copy, Clone, Debug, Serialize, Deserialize, PartialEq)]
+enum IgvmHashType {
+    Invalid = 0,
+    Sha256,
+    Sha384,
+    Sha512,
+}
+
+#[repr(u32)]
+#[derive(Copy, Clone, Debug, Serialize, Deserialize, PartialEq)]
+enum IgvmReportType {
+    Invalid = 0,
+    VbsVm,
+    SnpVm,
+    Tvm,
+    TdxVm,
+}
+
+#[repr(C)]
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+struct IgvmRequestData {
+    // Overall size of payload.
+    data_size: u32,
+    // Version of this structure, currently IGVM_ATTEST_VERSION_CURRENT (1).
+    version: u32,
+    report_type: IgvmReportType,
+    // The hash used in Report Data.
+    report_data_hash_type: IgvmHashType,
+    // Size of data blob.
+    variable_data_size: u32,
+    // Data holds AkPub or Transport Key.
+    variable_data: [u8; 0],
+}
+
+#[repr(u32)]
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+enum IgvmRequestType {
+    InvalidRequest = 0,
+    KeyReleaseRequest,
+    AkCertRequest,
+    KeyRotationInfoRequest,
+    KeyRotationStatus,
+}
+
+#[repr(C)]
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+struct AttestationHeader {
+    signature: u32,
+    version: u32,
+    report_size: u32,
+    request_type: IgvmRequestType,
+    status: u32,
+    reserved: [u32; 3],
+}
+
+#[repr(C)]
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+struct HwReport {
+    tdx_vm_report: TdxVmReport,
+    #[serde(with = "BigArray")]
+    _padding: [u8; MAX_REPORT_SIZE - MIN_REPORT_SIZE],
+}
+
+#[repr(C)]
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+struct AttestationReport {
+    header: AttestationHeader,
+    hw_report: HwReport,
+    hcl_data: IgvmRequestData,
+}
+
+pub struct HclReport {
+    bytes: Vec<u8>,
+    attestation_report: AttestationReport,
+    report_type: ReportType,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum ReportType {
+    Tdx,
+    Snp,
+}
+
+impl HclReport {
+    pub fn new(bytes: Vec<u8>) -> Result<Self, HclError> {
+        let attestation_report: AttestationReport = bincode::deserialize(&bytes)?;
+        let report_type = match attestation_report.hcl_data.report_type {
+            IgvmReportType::TdxVm => ReportType::Tdx,
+            IgvmReportType::SnpVm => ReportType::Snp,
+            _ => return Err(HclError::InvalidReportType),
+        };
+
+        let report = Self {
+            bytes,
+            attestation_report,
+            report_type,
+        };
+        Ok(report)
+    }
+
+    pub fn report_type(&self) -> ReportType {
+        self.report_type
+    }
+
+    pub fn tdx_report_slice(&self) -> &[u8] {
+        let tdx_report_offset = offset_of!(AttestationReport, hw_report);
+        let tdx_report_end = tdx_report_offset + mem::size_of::<TdxVmReport>();
```

Comment:
This is generally done to get rid of the padding, right?

Reply:
yeah there is a bit of padding (SNP reports are larger) but we also need to account for `hcl_data` in AttestationReport, as the last field